prog/vm/nexus.rb

```diff
@@ -545,6 +551,44 @@ def before_run
     nap 30
   end
 
+  label def suspending
+    # :nocov:
+    begin
+      host.sshable.cmd("sudo systemctl stop #{q_vm}")
+    rescue Sshable::SshError => ex
+      raise unless /Failed to stop .* Unit .* not loaded\./.match?(ex.stderr)
+    end
+
+    begin
+      host.sshable.cmd("sudo systemctl stop #{q_vm}-dnsmasq")
+    rescue Sshable::SshError => ex
+      raise unless /Failed to stop .* Unit .* not loaded\./.match?(ex.stderr)
+    end
+    # :nocov:
+
+    VmHost.dataset.where(id: vm.vm_host_id).update(
+      used_cores: Sequel[:used_cores] - vm.cores,
+      used_hugepages_1g: Sequel[:used_hugepages_1g] - vm.mem_gib
+    )
+
+    decr_suspend
+    hop_suspended
+  end
+
+  label def suspended
+    when_unsuspend_set? do
+      host.sshable.cmd("sudo systemctl start #{q_vm} #{q_vm}-dnsmasq")
+      VmHost.dataset.where(id: vm.vm_host_id).update(
+        used_cores: Sequel[:used_cores] + vm.cores,
+        used_hugepages_1g: Sequel[:used_hugepages_1g] + vm.mem_gib
+      )
+      decr_unsuspend
+      hop_wait
+    end
+
+    nap 2**35
```

Comment:
Do we have to nap this long? It simply means the unsespending will happen manually. Considering the semaphore increment will be done by the operator, I'm OK with this. But, maybe we can reduce it?

Reply:
well, strange thing, last system routinely napped for a thousand years and was more event-driven. This system is actually regressing in reducing spurious polling-like traffic relative to its predecessor.

So it's odd it shows in one place, yet, we're kinda doing something wrong everywhere else where we could fall asleep for a long time.

Reply:
Maybe we should schedule the strand when semaphore is incremented. It would definitely push things more towards being event driven and would encourage longer naps.

Reply:
I like the idea of being more event driven, and incrementing semaphores could be one of these events. In this case, someone needs to log into the console already and increment the `unsuspend` semaphore. It's probably ok to then also `run` the strand to immediately perform the unsuspend action.

Reply:
> Maybe we should schedule the strand when semaphore is incremented. It would definitely push things more towards being event driven and would encourage longer naps.

don't we do that? It tweaks sch to now.

```
  def self.incr(strand_id, name)
    DB.transaction do
      Strand.dataset.where(id: strand_id).update(schedule: Sequel::CURRENT_TIMESTAMP)
      Semaphore.create_with_id(strand_id: strand_id, name: name)
    end
  end
  ```

Reply:
Ah cool. I did not realize that.